Bump Twilio C# helper library to version 5.x

acamino opened the pull request:
The most relevant change introduced besides bumping the helper library is the introduction of a helper to test the XML endpoints. `ShouldReturnXmlResult` relies on the `ContentResult` that is being served instead of a `TwiMLResult`.

## TODO
- [x] `RequestValidator` will be introduced in the next RC of the Twilio C# helper library.
- [x] `CallResource.Create` is not working. I introduced `CallResource.CreateAsync` which works in the context of an ASP.NET MVC application.

For some reason the next-gen library doesn't work with non US phone numbers.

/cc: @joliveros, @hortega

acamino committed:
Bump Twilio C# helper library to version 5.x

acamino committed:
Use CreateAsync instead of Create for CallResource

`CallResource.Create` freezes the application.
This commit changes the signature of MakePhoneCallAsync reflects the
intereface of CreateAsync to make it more intuitive.

acamino committed:
Fix the issue when reseting the form on successful calls

acamino committed:
Use the "Origin" header to get the origin of the request

acamino committed:
Fix production URI handler generation

The test suite was brittle, this was not working with Url action helpers.
Those was generating `null` for any input. The introduced changes fix this
behavior.

acamino committed:
Remove unused key from Web.config

joliveros commented on ClickToCall.Web.Tests/packages.config:
```diff
@@ -5,11 +5,11 @@
   <package id="Microsoft.Web.Infrastructure" version="1.0.0.0" targetFramework="net452" />
   <package id="Moq" version="4.2.1510.2205" targetFramework="net452" />
-  <package id="Newtonsoft.Json" version="6.0.4" targetFramework="net452" />
+  <package id="Newtonsoft.Json" version="8.0.1" targetFramework="net452" />
```
Perhaps this could be updated to Newtonsoft.Json 9.0.1 to make it consistent throughout the solution?

acamino replied:
No problem, I can do it.

joliveros commented on ClickToCall.Web/Services/RequestValidationService.cs:
```diff
@@ -10,16 +13,46 @@ public interface IRequestValidationService
+            if (context.Request.HttpMethod == "POST")
+            {
+                // Iterate through that sorted list of POST parameters, and append the variable name and value (with no delimiters) to the end of the URL string
+                var sortedKeys = context.Request.Form.AllKeys.OrderBy(k => k, StringComparer.Ordinal).ToList();
```
perhaps this could benefit from the use of `Select(..)`? But not necessary

acamino replied:
`Select` is the equivalent to a map function, it goes from something of type `a` to something to type `b`. In this case we're just sorting an array of strings, right?

Something interesting here is that `ToList` is completely unnecessary. `OrderedBy` will return an `IOrderedEnumerable` so you can use the `foreach` statement without additional transformations.

joliveros replied:
I see. Is there any way we could use `Select` to append the keys to `value`? Probably `Aggregate` would be better

As it is it's fine. I was just curious 😉

acamino replied:
Furthermore, this class will be wiped out on the next release of the Twilio library. 😉

acamino replied:
That's not how `Select` is intended to be used. The key idea on `Select` a.k.a. _map_ is that when you apply the transformation function to all the elements in the collection you'll end up with another collection with the same size. `Aggregate` a.k.a. _reduce_ will _fold_ the elements in the collection into something else. Take a look at the canonical example for `Aggregate`:

```csharp
var nums = { 1, 2, 3, 4 };
var result = nums.Aggregate((a, b) => b + a); // result will be 10 as you might expect.

// This could make more sense for this example, though.
var sum = nums.Sum();
```

Both expressions are equivalent.

joliveros commented on ClickToCall.Web/Controllers/CallCenterController.cs:
```diff
@@ -42,23 +44,28 @@ public ActionResult Call(CallViewModel callViewModel)
 
-            return $"{requestUrlScheme}://{domain}{urlAction}";
+            var origin = Request.Headers[OriginHeader];
+            return $"{origin}{urlAction}";
```
nice!

acamino replied:
Pull that info from the header was brillant, thank you!

joliveros replied:
No worries. If you ever need any help with csharp just ask. 😉 

lol

joliveros commented on ClickToCall.Web/Controllers/CallCenterController.cs:
```diff
@@ -29,7 +31,7 @@ public ActionResult Index()
         /// </summary>
         [HttpPost]
-        public ActionResult Call(CallViewModel callViewModel)
+        public async Task<ActionResult> Call(CallViewModel callViewModel)
```
Perhaps this could be renamed to `CallAsync` ? Someone who really knows their stuff mentioned this to me 😉

joliveros reviewed (commented):
Hi @acamino, just minor comments. Everything else looks good! 

Thanks!

acamino committed:
Introduce FluentMvcTesting.Extensions to test XML endpoints

acamino committed:
Bump Newtonsoft.Json dependency to version 9.0.1

acamino committed:
Update FluentMvcTesting.Extensions to add `ControllerPropertiesMock`

This version includes a controller properties mock utility to make it
easier to test controllers that use URL helper methods.

joliveros commented:
LGTM

Thank you @acamino!

joliveros commented:
You're gonna hate me.. I forgot to mention, this bit from the top of the readme 😭 

```html
<a href="https://www.twilio.com">
  <img src="https://static0.twilio.com/marketing/bundles/marketing/img/logos/wordmark-red.svg" alt="Twilio" width="250" />
</a>
```

acamino committed:
Update README to reflect the changes made in the application

This commit also introduces a Local.config.example to ease the development
process.

acamino commented:
@joliveros, no worries. The requested changes for the README was introduced. Furthermore, I took the liberty to add a `Local.config.example` to ease the development process, we already had implemented this idea in other tutorials with `.env.example` and IMHO it works well.

Of course the instructions in the README was updated accordingly.

joliveros commented on ClickToCall.Web/Local.config.example:
```diff
@@ -0,0 +1,5 @@
+<appSettings>
```
This definitely is a time saver! +1

joliveros commented on README.md:
````diff
@@ -10,41 +14,27 @@ Click-to-call enables your company to convert web traffic into phone calls with
-   [Twilio Account Settings](https://www.twilio.com/user/account/settings).
 
-    Scared about ```[your-ngrok-subdomain]```?. Don't be!, go to [Using ngrok](#ngrok) section
+2. Copy the `ClickToCall.Web/Local.config.example` file to `ClickToCall.Web/Local.config`, and edit it including your credentials for the Twilio API (found at https://www.twilio.com/console/account/settings). You will also need a [Twilio Number](https://www.twilio.com/console/phone-numbers/incoming).
````
Much smoother way of getting this example up and running. +1

joliveros reviewed (approved):
Awesome work @acamino

Always a pleasure to review (learn from) your code!

Cheers

smendes committed:
Update twilio-csharp to latest stable 5.0.0
